### DISCUSSION TITLE
Unable to wrap QueryClientProvider

### DISCUSSION THREAD

**Discussion by Hydrair** [TanStack/query#3298] [Q&A]
I want to use a single Wrapper Component in a library im providing. The wrapper should provide stuff like Context and QueryClient. The component looks like this:

```
import { QueryClient, QueryClientProvider } from "react-query";

const client = new QueryClient();

function Wrapper(props) {
    return (
        <QueryClientProvider client={client}>
            {props.children}
        </QueryClientProvider>
    );
}
```
When wrapping children with the `Wrapper` useQuery throws an error that no QueryClientProvider is set. In this case `App` uses the useQuery Hook from react-query.
App component:
```
import { useQuery } from "react-query";

export const App = () => {
    const data = useQuery("data", () =>
        fetch("https://jsonplaceholder.typicode.com/todos/1").then((res) =>
            res.json()
        )
    );
    if (data.isError || data.isLoading) return <>Error/Loading</>;
    return <>{data.data.title}</>;
};
```
main.tsx
```
import { Wrapper } from "my-lib";

ReactDOM.render(
    <React.StrictMode>
        <Wrapper>
            <App />
        </Wrapper>
    </React.StrictMode>,
    document.getElementById("root")
);
```

At first i thought you need a QueryClientProvider directly above an useQuery hook. But for example in Storybook you can define one QueryClient for all stories together.

Vite has an open issue for this https://github.com/vitejs/vite/pull/3753 but in this SO post (https://stackoverflow.com/questions/71020363/using-queryclientprovider-in-a-wrapper) it seemed to be machine related?

**Comment by TkDodo** [comment]
How is "the library" referencing react-query? Does it have it as a `peerDependency` ? It looks like you have two versions of react-query, both bringing in their own context

**Reply by Hydrair** [reply]
I had it in the normal `dependencies`. When i add it to `peerDependencies` it should take the react-query from the App, right?
But then i'm not able to use react-query while testing with Storybook

**Reply by TkDodo** [reply]
If you have it in `dependencies`, you will "ship" that version of react-query as part of that libraries' node-modules, and the app will then have its own version of RQ.

`peerDependency` means the user (app) will be required to install a compatible version of react-query for itself, so this is what you want.

> But then i'm not able to use react-query while testing with Storybook

That's what devDependencies are for. So you want a devDependency + a peerDependency, like pretty much every other library :) For example, react-query has `react` and `react-dom` as peerDependency + as devDependency (for testing and development). Otherwise, we would ship our own version of react, which wouldn't be so nice :)

**Reply by Hydrair** [reply]
Ahh, alright, now i understand how this works. Thank you!
So in the library, react-query is in both `dev` and `peer` Dependencies. The library is linked `yarn link my-lib` to the App. In the App react-query is in the `dependencies`.

But it also doesn't matter if i create the QueryClient in the App or in the library, does it?. I get the components in the dev-tools like `Wrapper -> QueryClientProvider2 -> Context.Provider -> Context.Provider -> App`. But as soon as i want to query, this error gets thrown:
```QueryClientProvider.js:26 Uncaught Error: No QueryClient set, use QueryClientProvider to set one```.
I tried it with a new directory, maybe the build is misconfigured.

**Reply by TkDodo** [reply]
> But it also doesn't matter if i create the QueryClient in the App or in the library, does it?

yes, that shouldn't matter.

not sure how `yarn link` works, I was just phrasing this from a perspective where you release a library and install it via npm

**Reply by Hydrair** [reply]
Oh boy, i finally solved it. I forgot to declare the peerDeps as external in the `vite.config.ts`. Now I can build an awesome wrapper. Thanks for your help and clarifying the whole dependency stuff! :D

**Reply by ZhengRui** [reply]
> > But it also doesn't matter if i create the QueryClient in the App or in the library, does it?
> 
> yes, that shouldn't matter.
> 
> not sure how `yarn link` works, I was just phrasing this from a perspective where you release a library and install it via npm

I am in a dilemma now. My situation is: I have a package that contains `useQuery`, but expecting `QueryClientProvider` to be provided in the nextjs App that imports the package. 

From my knowledge, when using yarn to add local package, there are three ways:
1. `yarn link` inside package, and `yarn link package-name` inside app. this simply creates a symlink to package folder in app's node_modules.
2. `yarn add link:relative_path_to_package`, other than create the symlink to package folder, it also installs package's dependencies in app. 
3. `yarn add file:relative_path_to_package`, not create a symlink but make a full copy of package folder into app's node_modules, as well as install package's dependencies in app.

So from 1->3, it becomes heavier. Also for 2&3, it ends up with two copies of package's dependencies, one inside package itself and one inside the app. 

I encountered `hooks (package used zustand's `useStore`) outside of function component` issue because I put `react` and `react-dom` in both `devDependencies` and `peerDependencies` of the package, so react gets installed inside the package, and the nextjs app also comes with react, so two versions of react caused the above issue. I removed `react` and `react-dom` from `devDependencies` of the package, so they are not included in the package node_modules, and then only the `react` from the app is used and the issue is then solved. ***But I have `@types/react` to save me so I can still build the package.***

However for `react-query`, its types is included in itself. So to be able to build package, it has to be installed. And because the `QueryClientProvider` is provided by the app, so it also has to be installed in the app. So two `react-query` causes the `no queryClient set` issue even with `contextSharing` set true in the provider in the app. 

I hope I made my situation clear, any suggestions @TkDodo ?

**Reply by ZhengRui** [reply]
I found a hacky solution:

in nextjs app `yarn add link:relative_path_to_react_query_in_package_node_modules` , in package set `react-query` as `dependencies`, this way the app is using the package's `react-query`.

Ideally I prefer the package uses app's `react-query` by setting it as `peerDependencies` and something like `@types/react-query` as `devDependencies` in the package, similar to `react` case.

**Reply by TkDodo** [reply]
> I have a package that contains useQuery, but expecting QueryClientProvider to be provided in the nextjs App that imports the package.

That's a bit of a leaky abstraction imo. How does the App know which keys the package is using? It's a bit like mixins where you can easily get conflicts.

**Reply by ZhengRui** [reply]
ahhh, i see your point. You mean if the provider is in the App, then there maybe key conflicts between the App and the package. Right, so if i am to write a react library that uses `useQuery`, i should also wrap the provider inside the library? I think i did it this way because unlike App, there isn't an entry component in the package, so my first thought was to put the provider outside of the package.

**Reply by TkDodo** [reply]
in v4 you can pass a custom context to achieve isolation:

- https://tanstack.com/query/v4/docs/react/guides/migrating-to-react-query-4#custom-contexts-for-multiple-providers

this will change in v5 though towards custom queryClient to be passed, which is independent of react

**Reply by ZhengRui** [reply]
Thanks for pointing out, have updated the package code and the way to use it in App is much cleaner now
